### PR TITLE
Add constructor for mocks

### PR DESCRIPTION
**Pull request by grongor**
These changes will simplify the tests by not being required to assert mocks expectations every time. This will also eliminate situations where your tests are not failing even though they should, only to find out that you forget the defer statement.

Example
Before:
```go
factory := &mocks.Factory{}
defer factory.AssertExpectations()

something := &mocks.Something{}
defer something.AssertExpectations()

something.On("GetTime").Return("tooLate")

factory.On("Create").Return(something)

testThings(factory)
```
Now:
```go
factory := mocks.NewFactory()
something := mocks.NewSomething()

something.On("GetTime").Return("tooLate")

factory.On("Create").Return(something)

testThings(factory)
```

edit: this already got merged: These changes are based on a branch where I did some refactoring, you can see it here https://github.com/vektra/mockery/pull/399. If you don't intend to merge the refactoring PR, I can rebase this branch onto the master.

### CODE REVIEW

**Comment by LandonTClipp** [comment]
@grongor  can you please rebase onto master and fix the conflicts? Thanks!

**Comment by grongor** [comment]
@LandonTClipp done :)

File: pkg/generator.go
Comment (LandonTClipp) [inline pkg/generator.go:572]: Would it be possible you could use golang templates for this? This is kind of difficult to read and I think the templates would make it nicer.

Comment (grongor) [reply pkg/generator.go:572]: Well I could do that but to be honest I don't like it :D 
```
diff --git a/pkg/generator.go b/pkg/generator.go
index f26c50c..b969a6a 100644
--- a/pkg/generator.go
+++ b/pkg/generator.go
@@ -8,6 +8,7 @@ import (
 	"go/ast"
 	"go/build"
 	"go/types"
+	"html/template"
 	"io"
 	"os"
 	"path/filepath"
@@ -569,9 +570,9 @@ func (g *Generator) Generate(ctx context.Context) error {
 
 func (g *Generator) generateConstructor() {
 	const constructor = `
-// %[1]sew%[2]s creates a new instance of %[3]s. It also registers a cleanup function to assert the mocks expectations.
-func %[1]sew%[2]s(t testing.TB) *%[3]s {
-	mock := &%[3]s{}
+// {{.FuncFirstLetter}}ew{{.MockNameUpperCase}} creates a new instance of {{.MockName}}. It also registers a cleanup function to assert the mocks expectations.
+func {{.FuncFirstLetter}}ew{{.MockNameUpperCase}}(t testing.TB) *{{.MockName}} {
+	mock := &{{.MockName}}{}
 
 	t.Cleanup(func() { mock.AssertExpectations(t) })
 
@@ -592,7 +593,19 @@ func %[1]sew%[2]s(t testing.TB) *%[3]s {
 		break
 	}
 
-	g.printf(constructor, funcFirstLetter, mockNameUpperCase, mockName)
+	err := template.Must(template.New("").Parse(constructor)).Execute(
+		&g.buf,
+		struct {
+			FuncFirstLetter, MockNameUpperCase, MockName string
+		}{
+			FuncFirstLetter:   funcFirstLetter,
+			MockNameUpperCase: mockNameUpperCase,
+			MockName:          mockName,
+		},
+	)
+	if err != nil {
+		panic("failed to write mock constructor: " + err.Error())
+	}
 }
```

Just now I realized I can simplify the "printf" version, what do you think?
```
diff --git a/pkg/generator.go b/pkg/generator.go
index f26c50c..5deb823 100644
--- a/pkg/generator.go
+++ b/pkg/generator.go
@@ -569,9 +569,9 @@ func (g *Generator) Generate(ctx context.Context) error {
 
 func (g *Generator) generateConstructor() {
 	const constructor = `
-// %[1]sew%[2]s creates a new instance of %[3]s. It also registers a cleanup function to assert the mocks expectations.
-func %[1]sew%[2]s(t testing.TB) *%[3]s {
-	mock := &%[3]s{}
+// %[1]s creates a new instance of %[2]s. It also registers a cleanup function to assert the mocks expectations.
+func %[1]s(t testing.TB) *%[2]s {
+	mock := &%[2]s{}
 
 	t.Cleanup(func() { mock.AssertExpectations(t) })
 
@@ -592,7 +592,7 @@ func %[1]sew%[2]s(t testing.TB) *%[3]s {
 		break
 	}
 
-	g.printf(constructor, funcFirstLetter, mockNameUpperCase, mockName)
+	g.printf(constructor, funcFirstLetter+"ew"+mockNameUpperCase, mockName)
 }

```

Comment (grongor) [reply pkg/generator.go:572]: @LandonTClipp Hi, can we please finish this? I'd really love to have the constructors for all my mocks :)

**Comment by LandonTClipp** [comment]
@grongor I tried fixing a merge conflict but messed up a closing bracket, could you fix that? I will merge once the tests pass.

**Comment by grongor** [comment]
@LandonTClipp Done :). Also please note that the Expecter PR did more changes than it should (refactoring, new way to check generated code, ...), specifically `expectedBytes, err := ioutil.ReadFile(getFixturePath("mocks", "expecter.go"))`. I'm not saying that it is bad...just that this is the first place in the codebase that it is done this way, so I had to look around for some time until I managed to find a way to fix the errors after rebasing. I also reverted part of the Expecter PR because it caused panics during tests (it's in a separate commit, please check).

**Comment by LandonTClipp** [comment]
Thanks @grongor, can you elaborate on what you mean by "new way to check generated code"?

File: pkg/generator.go
Comment (LandonTClipp) [inline pkg/generator.go:723]: I've seen a repeated pattern of where we manually modify the first letter depending on whether it's exported or not. Might make sense have a helper function that will make a string exported/non-exported depending on the state of some other variable/function name.

File: Makefile
Comment (LandonTClipp) [inline Makefile:4]: 👍🏻

File: pkg/generator_test.go
Comment (LandonTClipp) [inline pkg/generator_test.go:86]: 👍🏻 Thanks for catching this.

**Review by LandonTClipp — APPROVED** [review]
PR looks good, I only have one comment about the export/unexported name logic but that doesn't hold up this PR. If you want extra credit you can make the helper function :)

**Comment by LandonTClipp** [comment]
Because we're dealing with some issues regarding go1.18 builds I'm going to pause this PR to minimize changes, but once that settles down this is good to go.

**Comment by grongor** [comment]
> Can you elaborate on what you mean by "new way to check generated code"?

this file: `pkg/fixtures/mocks/expecter.go`

The rest of the generated mock files live in `mocks` directory. The generator test contains expected code directly in the test file `pkg/generator_test.go`, and there is only one test using an "external" file `pkg/fixtures/mocks/expecter.go`. Maybe there is a good reason for it - I don't know. I just wanted to point out that it kind of does things differently than the rest of the tests.

> Because we're dealing with some issues...

No problem, I'm glad that it's finally moving :) Thanks!

File: pkg/generator.go
Comment (grongor) [reply pkg/generator.go:723]: I didn't but yeah if you know about such places it would make sense to make a helper function :) I know about the `upperFirstOnly`, but it always changes the string (doesn't check another).

Comment (LandonTClipp) [reply pkg/generator.go:723]: If you could create a helper function in a new pkg/strings.go or whatever name makes sense, that would be great. I'm sure we'll use it again. Thanks!

**Comment by LandonTClipp** [comment]
@grongor merged and deployed. One additional matter, could you document this in the README as a new feature that is available? I'm sure others will benefit from it.

**Comment by grongor** [comment]
@LandonTClipp thank you! I'll look into updating the README soon, as well as the helper function you mentioned (https://github.com/vektra/mockery/pull/406#discussion_r840819968).

**Comment by grongor** [comment]
https://github.com/vektra/mockery/pull/450

File: pkg/generator.go
Comment (grongor) [reply pkg/generator.go:723]: See this: https://github.com/vektra/mockery/pull/452

**Pull request merged**
